modules/compression/src/main/java/org/apache/ignite/spi/communication/ZstdEngine.java

```diff
@@ -0,0 +1,235 @@
+                        compressArr = new byte[compressArr.length * 2];
+                    else
+                        throw new IOException("Failed to compress data: " + Zstd.getErrorName(res));
+                }
+            }
+            while (res == DEST_BUFFER_OVERFLOW_ERR);
+
+            if (res > buf.remaining()) {
+                src.rewind();
+
+                return BUFFER_OVERFLOW;
+            }
+
+            buf.put(compressArr, 0, (int)res);
+
+            return OK;
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompressionEngineResult decompress(ByteBuffer src, ByteBuffer buf) throws IOException {
+        int frameSize = readFrameSize(src);
+
+        if (frameSize == -1)
+            return BUFFER_UNDERFLOW;
+
+        if (src.isDirect() && buf.isDirect()) {
+            int oldLimit = src.limit();
+
+            src.limit(src.position() + frameSize);
+
+            try {
+                Zstd.decompress(buf, src);
+            }
+            catch (RuntimeException e) {
+                src.limit(oldLimit);
+
+                if (e.getMessage().contains(DEST_BUFFER_OVERFLOW_ERR_MSG))
+                    return BUFFER_OVERFLOW;
+                else
+                    throw new IOException("Failed to compress data", e);
+            }
+
+            src.limit(oldLimit);
+
+            return OK;
+        }
+        else {
+            int initPos = src.position();
+
+            byte[] inputWrapArray = new byte[frameSize];
+
+            src.get(inputWrapArray);
+
+            long res;
+
+            do {
+                res = Zstd.decompress(decompressArr, inputWrapArray);
+
+                if (Zstd.isError(res)) {
+                    if (res == DEST_BUFFER_OVERFLOW_ERR)
+                        decompressArr = new byte[decompressArr.length * 2];
+                    else
+                        throw new IOException("Failed to decompress data: " + Zstd.getErrorName(res));
+                }
+            }
+            while (res == DEST_BUFFER_OVERFLOW_ERR);
+
+            if (res > buf.remaining()) {
+                src.position(initPos);
+
+                return BUFFER_OVERFLOW;
+            }
+
+            buf.put(decompressArr, 0, (int)res);
+
+            return OK;
+        }
+    }
+
+    /**
+     * Read size of compressed frame.
+     *
+     * @param buf ByteBuffer.
+     * @return size of compressed frame, -1 otherwise if not enough data.
+     */
+    private static int readFrameSize(ByteBuffer buf) throws IOException {
+        int initPos = buf.position();
+        int limit = buf.limit();
+
+        if (initPos + 4 + 1 + 3 /* MAGIC + FRAME_HEADER + BLOCK_HEADER */ > limit)
+            return -1;
+
+        int offset = initPos;
+
+        boolean magicCheck = buf.get(offset++) == 40 &&
+                            buf.get(offset++) == -75 &&
+                            buf.get(offset++) == 47 &&
+                            buf.get(offset++) == -3;
```

Comment:
I think it's better to describe all of this magic number in some `static final` field.